eth/executionclient/execution_client.go

```diff
@@ -31,76 +32,164 @@ var (
-	err := client.connect(ctx)
+
+	// Initialize ManagedClients
+	for _, addr := range client.nodeAddrs {
+		mc, err := NewManagedClient(ctx, addr, client.logger, client.reconnectionInitialInterval, client.reconnectionMaxInterval, client.syncDistanceTolerance)
+		if err != nil {
+			client.logger.Error("Failed to initialize ManagedClient", zap.String("address", addr), zap.Error(err))
+			return nil, fmt.Errorf("create managed client: %w", err)
+		}
+		client.clients = append(client.clients, mc)
+	}
+
+	if len(client.clients) == 0 {
+		return nil, fmt.Errorf("no execution clients could be initialized")
+	}
+
+	same, err := client.assertSameChainIDs(ctx)
 	if err != nil {
-		return nil, fmt.Errorf("failed to connect to execution client: %w", err)
+		return nil, fmt.Errorf("assert same chain IDs: %w", err)
+	}
+	if !same {
+		return nil, fmt.Errorf("execution clients' chain IDs are not same")
 	}
 
 	client.syncProgressFn = client.syncProgress
 
 	return client, nil
 }
 
-func (ec *ExecutionClient) syncProgress(ctx context.Context) (*ethereum.SyncProgress, error) {
-	return ec.client.SyncProgress(ctx)
+// assertSameChainIDs checks if all healthy clients have the same chain ID.
+// It sets firstChainID to the chain ID of the first healthy client encountered.
+func (ec *ExecutionClient) assertSameChainIDs(ctx context.Context) (bool, error) {
+	ec.clientsMu.RLock()
+	defer ec.clientsMu.RUnlock()
+
+	var firstChainID *big.Int
+	for _, client := range ec.clients {
+		c := client.getClient()
+		if c == nil {
+			ec.logger.Warn("Client is not healthy", zap.String("address", client.addr))
+			return false, fmt.Errorf("client is not healthy")
+		}
+		chainID, err := c.ChainID(ctx)
+		if err != nil {
+			ec.logger.Error("Failed to get chain ID", zap.String("address", client.addr), zap.Error(err))
+			return false, fmt.Errorf("get chain ID: %w", err)
+		}
+		if firstChainID == nil {
+			firstChainID = chainID
+			continue
+		}
+		if firstChainID.Cmp(chainID) != 0 {
+			ec.logger.Error("Chain ID mismatch",
+				zap.String("first_chain_id", firstChainID.String()),
+				zap.String("current_chain_id", chainID.String()),
+				zap.String("address", client.addr))
+			return false, nil
+		}
+	}
+
+	if firstChainID == nil {
+		return false, fmt.Errorf("no healthy clients to determine chain ID")
+	}
+
+	return true, nil
+}
+
+// syncProgress retrieves the sync progress using a healthy client.
+func (ec *ExecutionClient) syncProgress(ctx context.Context, mc *ManagedClient) (*ethereum.SyncProgress, error) {
+	client := mc.getClient()
+	if client == nil {
+		return nil, ErrNotConnected
+	}
+	sp, err := client.SyncProgress(ctx)
+	if err != nil {
+		ec.logger.Error(elResponseErrMsg,
+			zap.String("method", "eth_syncing"),
+			zap.Error(err))
+		return nil, err
+	}
+	return sp, nil
 }
 
 // Close shuts down ExecutionClient.
 func (ec *ExecutionClient) Close() error {
-	close(ec.closed)
-	ec.client.Close()
+	ec.closingOnce.Do(func() {
+		close(ec.closed)
+		ec.clientsMu.RLock()
```

Comment:
IMO this seems like too much concurrency primitives to do a simple thing. maybe we can just use context?

Reply:
fixed